crates/flipperzero/src/furi/stream_buffer.rs

```diff
@@ -0,0 +1,412 @@
+        unsafe {
+            Self(NonNull::new_unchecked(sys::furi_stream_buffer_alloc(
+                size.into(),
+                trigger_level,
+            )))
+        }
+    }
+
+    fn set_trigger_level(&self, trigger_level: usize) -> EmptyResult {
+        let updated = unsafe { sys::furi_stream_set_trigger_level(self.0.as_ptr(), trigger_level) };
+        match updated {
+            true => Ok(()),
+            false => Err(()),
+        }
+    }
+
+    fn send(&self, data: &[u8], timeout: u32) -> usize {
+        unsafe {
+            sys::furi_stream_buffer_send(self.0.as_ptr(), data.as_ptr().cast(), data.len(), timeout)
+        }
+    }
+
+    fn receive(&self, data: &mut [u8], timeout: u32) -> usize {
+        unsafe {
+            sys::furi_stream_buffer_receive(
+                self.0.as_ptr(),
+                data.as_mut_ptr().cast(),
+                data.len(),
+                timeout,
+            )
+        }
+    }
+
+    fn bytes_available(&self) -> usize {
+        unsafe { sys::furi_stream_buffer_bytes_available(self.0.as_ptr()) }
+    }
+
+    fn spaces_available(&self) -> usize {
+        unsafe { sys::furi_stream_buffer_spaces_available(self.0.as_ptr()) }
+    }
+
+    fn is_full(&self) -> bool {
+        unsafe { sys::furi_stream_buffer_is_full(self.0.as_ptr()) }
+    }
+
+    fn is_empty(&self) -> bool {
+        unsafe { sys::furi_stream_buffer_is_empty(self.0.as_ptr()) }
+    }
+
+    fn reset(&self) -> EmptyResult {
+        let status = unsafe { sys::furi_stream_buffer_reset(self.0.as_ptr()) };
+        let status = Status(status);
+        match status {
+            Status::OK => Ok(()),
+            Status::ERR => Err(()),
+            _ => unreachable!("furi_stream_buffer_reset only returns Ok or Error"),
+        }
+    }
+}
+
+impl Drop for StreamBuffer {
+    fn drop(&mut self) {
+        // SAFETY:
+        // Since we keep an Arc in both, the sender and receiver, we know we only drop when both of
+        // them are dropped too.
+        unsafe {
+            sys::furi_stream_buffer_free(self.0.as_ptr());
+        }
+    }
+}
+
+/// Sender side of a furi stream buffer.
+///
+/// An instance can be obtained using the [`stream_buffer`] function.
+///
+/// This struct allows sending data through the stream buffer, checking its state, or resetting it.
+/// Use the [`is_receiver_alive`](Self::is_receiver_alive) method to verify if the associated
+/// [`Receiver`] is still alive, helping to avoid sending data that will not be read.
+pub struct Sender {
+    buffer_ref: Arc<StreamBuffer>,
+}
+
+impl Debug for Sender {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        // the receiver_alive field is not real but a nice debug information
+        f.debug_struct("Sender")
+            .field("buffer_ref", &self.buffer_ref)
+            .field("receiver_alive", &self.is_receiver_alive())
+            .finish()
+    }
+}
+
+unsafe impl Send for Sender {}
```

Comment:
Safety comment would be useful here

Reply:
I don't know what to really say here. Usually you would argue why `Send` is sound at this point but that's not really the reason that this type implements `Send`. Maybe I need to write quite some paragraph about the `Send` here, just like in `StreamBuffer`.

Not sure.

Reply:
I made `StreamBuffer` `Send` and `Sync` with some comments why that is safe. For that I marked `send` and `receive` of the `StreamBuffer´  as `unsafe`.